### PR TITLE
Alex/create review page

### PR DESCRIPTION
**Pull request by uyxela**
## Notion ticket link

[Create/Edit Review Base Page](https://www.notion.so/uwblueprintexecs/Create-Edit-Review-Base-Page-0c7d7a44c1534085b9d69a79a5a53d67)

Screen Shot 2022-03-02 at 2 57 15 AM
Screen Shot 2022-03-02 at 2 58 24 AM

## Implementation description
* Page for creating new reviews
* Tab bar at top of page with buttons for saving, publishing, deleting the review
* Add new book button and shows list of books
* Review textbox (uses React Quill)
* Review byline
* Featured radio buttons
* Several modals for different actions (deleting book, deleting review, publishing review)

## Steps to test
1. Page path is `/create-review`
2. Interact with page elements

## Notes
* Originally branched off of `christine/review-book-modal`
* Created a new branch and cherry-picked commits 
* Made several additional changes (rearrange files, comment out components) to mitigate merge conflicts

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have run docker-compose up and my project compiled
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: frontend/src/types/BookTypes.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+export type Publisher = {
+  fullName: string;
+  publishYear: number;
+};
+
+export type Book = {
+  title: string;
+  coverImage: string;
+  titlePrefix: string;
+  seriesOrder: string;
```
Comment (susheelk) [inline frontend/src/types/BookTypes.ts:22]: I think this should be an int. Could you verify?

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+// import { Option } from "../../../types/TagTypes";
+// import BookModal from "./BookModal";
+import DeleteModal from "./DeleteBookModal";
+import DeleteReviewModal from "./DeleteReviewModal";
+import PublishModal from "./PublishModal";
+import ReviewEditor from "./ReviewEditor";
+import SingleBook from "./SingleBook";
+
+const CreateReview = (): React.ReactElement => {
+  // const [tagsSelected, setTagsSelected] = useState<Option[]>([]);
```
Comment (susheelk) [inline frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:26]: Are these commented lines necessary? If so, could you add a comment detailing why?

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+import { Book } from "../../../types/BookTypes";
+// import { Option } from "../../../types/TagTypes";
+// import BookModal from "./BookModal";
+import DeleteModal from "./DeleteBookModal";
+import DeleteReviewModal from "./DeleteReviewModal";
+import PublishModal from "./PublishModal";
+import ReviewEditor from "./ReviewEditor";
+import SingleBook from "./SingleBook";
+
+const CreateReview = (): React.ReactElement => {
```
Comment (susheelk) [inline frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:25]: Could you add a header describing this component?

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const editBook = (book: Book, index: number) => {
+    deleteBook(index);
+    addBook(book);
+  };
+
+  // useEffect hook adds dummy data to the books array
+  useEffect(() => {
+    setBooks([
+      {
```
Comment (susheelk) [inline frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:70]: Would it be better to move the mock data to a `const` outside of here?

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+          },
+        ],
+        seriesName: "The Amazing Series",
+      },
+    ]);
+  }, []);
+
+  return (
+    <Box>
+      {/* <BookModal
```
Comment (susheelk) [inline frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:107]: Do we know for sure this is what the props for the BookModal will look like? Otherwise it might make more sense to remove this for now and let @liu-jeffrey add this later

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+            cursor="pointer"
+            m={6}
+            ml={0}
+            // onClick={() => setShowBookModal(true)}
+          >
+            <IconButton
+              aria-label="Add new book"
+              variant="outline"
+              h="180px"
+              // should this be 112px for a 1.6:1 ratio?
```
Comment (susheelk) [inline frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:219]: Good question, could you check with a designer?

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+          display="flex"
+          flexDirection="row"
+          alignItems="flex-start"
+          justifyContent="flex-start"
+          mt="10px"
+          mb="50px"
+        >
+          {/* Current books display */}
+          {books.map((book, i) => (
+            <SingleBook
```
Comment (susheelk) [inline frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:196]: Theoretically, these should be displayed sorted by `seriesOrder`. That might be better done in another PR though, what do you think?

File: frontend/src/components/pages/CreateReview/DeleteReviewModal.tsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+} from "@chakra-ui/react";
+import React from "react";
+
+interface DeleteReviewModalProps {
+  isOpen: boolean;
+  onClose: () => void;
+  deleteReview: () => void;
+}
+
+const DeleteReviewModal = (
```
Comment (susheelk) [inline frontend/src/components/pages/CreateReview/DeleteReviewModal.tsx:20]: Could you add a docstring header to all these components?

File: frontend/src/components/pages/CreateReview/ReviewEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  setValue,
+}: {
+  value: string;
+  setValue: React.Dispatch<React.SetStateAction<string>>;
+}): React.ReactElement => {
+  const modules = {
+    toolbar: [
+      ["bold", "italic", "underline", "strike"],
+      [{ header: [1, 2, 3, false] }],
+      ["link", "blockquote", "image", { list: "bullet" }, { list: "ordered" }],
```
Comment (susheelk) [inline frontend/src/components/pages/CreateReview/ReviewEditor.tsx:18]: Could you check with a designer if we need images here?

File: frontend/src/components/pages/CreateReview/SingleBook.tsx [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+import { Book } from "../../../types/BookTypes";
+
+interface SingleBookProps {
+  book: Book;
+  index: number;
+  showModal: (showDeleteBookModal: boolean) => void;
+  setIndex: (deleteBookIndex: number) => void;
+}
+
+const SingleBook = (props: SingleBookProps): React.ReactElement => {
```
Comment (susheelk) [inline frontend/src/components/pages/CreateReview/SingleBook.tsx:14]: Docstring header here too please

File: frontend/src/components/pages/CreateReview/editor.css [hunk 11]
```diff
@@ -0,0 +1,947 @@
+/*!
```
Comment (susheelk) [inline frontend/src/components/pages/CreateReview/editor.css:1]: We should maybe move all of our styling to another folder (ie `frontend/styles/**` or something). What do you think?

File: frontend/yarn.lock [hunk 12]
```diff
@@ -1123,7 +1123,7 @@
 
 "@babel/runtime-corejs3@^7.12.1":
   version "7.16.3"
-  resolved "https://registry.yarnpkg.com/@babel/runtime-corejs3/-/runtime-corejs3-7.16.3.tgz#1e25de4fa994c57c18e5fdda6cc810dac70f5590"
```
Comment (susheelk) [inline frontend/yarn.lock:1126]: Do you know why these hashes show up here? Might cause some conflicts with other branches

**Review by susheelk — CHANGES_REQUESTED** [review]

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+          <IconButton
+            aria-label="Go back"
+            variant="ghost"
+            isRound
+            icon={<ArrowBackIcon w={7} h={7} />}
+            mr={6}
+            onClick={() => window.history.back()}
+          />
+          <Text fontSize="18px" fontWeight="semibold">
+            Back to Dashboard
```
Comment (susheelk) [inline frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:156]: I think this should be a link back as well

**Review by susheelk — CHANGES_REQUESTED** [review]

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+  const onDeleteBookModalClose = () => setShowDeleteBookModal(false);
+  const onPublishModalClose = () => setShowPublishModal(false);
+  const onDeleteReviewModalClose = () => setShowDeleteReviewModal(false);
+
+  // const handleTagSelected = (e: Option[]) => {
+  //   setTagsSelected(e);
+  // };
+
+  // adds a book to the list of books
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (liu-jeffrey) [inline frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:50]: can you get rid of this es-lint disable? i think all the variables are used here

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx (hunk 4 above)
Comment (liu-jeffrey) [reply frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:70]: +1

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+        deleteReview={() => {}}
+      />
+      {/* Tool bar */}
+      <Box
+        display="flex"
+        flexDirection="row"
+        alignItems="center"
+        justifyContent="space-between"
+        m="0"
+        mb="60px"
```
Comment (liu-jeffrey) [inline frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:138]: Is it possible to combine these two into one prop? I'm not super familiar with Chakra for this case so I could be wrong.

File: frontend/src/components/pages/CreateReview/PublishModal.tsx [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+            <Button
+              colorScheme="teal"
+              bgColor="#0EBCBD"
+              w="48%"
+              onClick={() => {
+                publishBook();
+                onClose();
+              }}
+            >
+              Yes, publish book
```
Comment (liu-jeffrey) [inline frontend/src/components/pages/CreateReview/PublishModal.tsx:57]: Should this be publish review* ?

**Review by liu-jeffrey — DISMISSED** [review]
Solid work Alex! Was wondering if you could also attach some screenshots in your PR? Thank you!

**Review by mahadzaryab1 — COMMENTED** [review]
Nice work - added a few comments. Also, start a thread on the slack channel with the designers for them to do a design review.

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+  Radio,
+  RadioGroup,
+  Stack,
+  Text,
+} from "@chakra-ui/react";
+import React, { useEffect, useState } from "react";
+
+import { Book } from "../../../types/BookTypes";
+// import { Option } from "../../../types/TagTypes";
+// import BookModal from "./BookModal";
```
[the remark this answers is not in the document]
Comment (liu-jeffrey) [reply frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:18]: I need the BookModal for my PR that's based on this PR. We can remove them if you want though

[the remark this answers is not in the document]
Comment (mahadzaryab1) [reply frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:18]: no worries - just wanted to check!

**Review by susheelk — APPROVED** [review]

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 18]
```diff
@@ -152,7 +132,12 @@ const CreateReview = (): React.ReactElement => {
             mr={6}
             onClick={() => window.history.back()}
           />
-          <Text fontSize="18px" fontWeight="semibold">
+          <Text
+            fontSize="18px"
+            fontWeight="semibold"
+            onClick={() => window.history.back()}
```
Comment (mahadzaryab1) [inline frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:138]: What if someone lands directly on this page?

File: frontend/src/components/pages/CreateReview/CreateReviewPage.tsx [hunk 19]
```diff
@@ -191,16 +176,26 @@ const CreateReview = (): React.ReactElement => {
...
+          {/* Current books display (sorted by seriesOrder) */}
+          {books
+            .sort((a: Book, b: Book) => {
+              if (a.seriesOrder < b.seriesOrder) {
+                return -1;
+              }
+              if (a.seriesOrder > b.seriesOrder) {
+                return 1;
+              }
+              return 0;
```
Comment (mahadzaryab1) [inline frontend/src/components/pages/CreateReview/CreateReviewPage.tsx:188]: could use `if...else if...else` here

**Review by mahadzaryab1 — COMMENTED** [review]
a couple more small comments
